### PR TITLE
Eng 9458 basic functional testing

### PR DESCRIPTION
**Pull request by jacobweinstock**
## Description

This PR adds a foundation for functional/end to end testing. 

## Why is this needed

Code confidence and quality need to be improved. This PR tries to address those.

Fixes: #

## How Has This Been Tested?
See `test/README.md`.  add resources to file: `test/resources.yaml` then run `make test-functional`

## Checklist:

I have:

- [X] updated the documentation and/or roadmap (if required)
- [X] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: test/machine/main.go [hunk 1]
```diff
@@ -15,12 +16,12 @@ import (
 	grpc_zap "github.com/grpc-ecosystem/go-grpc-middleware/logging/zap"
 	grpc_ctxtags "github.com/grpc-ecosystem/go-grpc-middleware/tags"
 	grpc_validator "github.com/grpc-ecosystem/go-grpc-middleware/validator"
+	packet_logr "github.com/packethost/pkg/log/logr"
```
Comment (mmlb) [inline test/machine/main.go]: logr?

File: test/machine/main.go [hunk 2]
```diff
@@ -136,3 +158,14 @@ func captureOutput(f func()) string {
 	writer.Close()
 	return <-out
 }
+
+// MemorySink implements zap.Sink by writing all messages to a buffer.
```
Comment (mmlb) [inline test/machine/main.go]: hmm iirc there's a zap debug pkg that might be better?

File: api/v1/task.proto [hunk 3]
```diff
@@ -35,5 +35,5 @@ message Error {
 
     // Additional error information that the client code can use to handle
     // the error, such as retry delay or a help link.
-    repeated google.protobuf.Any details = 3;
+    repeated string details = 3;
```
Comment (mmlb) [inline api/v1/task.proto:38]: why this change? This should probably be best on its own commit.

File: cmd/server.go [hunk 4]
```diff
@@ -74,7 +74,7 @@ var serverCmd = &cobra.Command{
 		*/
```
Comment (mmlb) [inline cmd/server.go:74]: Not a good place to put this comment, but can you add some context to the commit message. Something along the lines of high level of how the tests are supposed to work.

File: test/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+```bash
+make test-functional
+# under the hood it calls: go test -v ./test/... --tags=functional -config 'resources.yaml'
+```
+
+## Development
+
+* The functional tests use the standard `Go` testing library.
+* Test files should follow normal `go test` patterns and be named with `_test.go`
+* Test files should contain a [build tag](https://golang.org/cmd/go/#hdr-Build_constraints) of `// +build functional` at the start of the file.
```
Comment (mmlb) [inline test/README.md:48]: why?

File: .gitignore [hunk 6]
```diff
@@ -4,4 +4,4 @@ pbnj
 bin/
 pid
 temp.in
-test/resource.yaml
\ No newline at end of file
+test/resources.yaml
```
Comment (mmlb) [inline .gitignore:7]: new line

**Review by mmlb — CHANGES_REQUESTED** [review]
Looks good, I think the intermediate drafts of test setup should all be squashed.

File: test/machine/main.go (hunk 1 above)
Comment (jacobweinstock) [reply test/machine/main.go]: sorry, bad rebase. should be good now.

File: test/machine/main.go (hunk 2 above)
Comment (jacobweinstock) [reply test/machine/main.go]: sorry, bad rebase. this file should be removed now

File: test/main.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	port := "40041"
+
+	go func() {
+		runGrpcServer(port)
+		//serverCmd := cmd.NewRootCmd()
+		//serverCmd.SetArgs([]string{"server", "--port", port})
+		//serverCmd.Execute()
+	}()
+
+	time.Sleep(2 * time.Second)
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply test/main.go]: sorry, bad rebase. this file should be remove now.

File: api/v1/task.proto (hunk 3 above)
Comment (jacobweinstock) [reply api/v1/task.proto:38]: made it harder to test as a `google.protobuf.Any`

File: pkg/oob/oob.go [hunk 8]
```diff
@@ -23,7 +19,12 @@ type BMC interface {
...
+// Error for all bmc actions
+// type Error v1.Error
+
 // Error for all bmc actions
 type Error struct {
-	v1.Error
+	Code    int32
+	Message string
+	Details []string
 }
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply pkg/oob/oob.go:30]: sure, i'll update

File: test/README.md (hunk 5 above)
Comment (jacobweinstock) [reply test/README.md:48]: so that `go test ./...` doesn't run the functional tests

File: .gitignore (hunk 6 above)
Comment (jacobweinstock) [reply .gitignore:7]: will update

File: cmd/server.go (hunk 4 above)
Comment (jacobweinstock) [reply cmd/server.go:74]: tests? sorry, not following. I can put this comment showing use of a different persistence layer in a doc tho.

Comment (jacobweinstock) [reply cmd/server.go]: fyi, removed comments

File: server/grpcsvr/taskrunner/taskrunner.go [hunk 9]
```diff
@@ -11,9 +11,7 @@ import (
 	"github.com/pkg/errors"
```
Comment (mmlb) [inline server/grpcsvr/taskrunner/taskrunner.go:12]: should drop this extra new-line

File: test/README.md (hunk 5 above)
Comment (mmlb) [reply test/README.md:48]: :facepalm: I thought we were also faking the BMC accesses so thought we could do functional tests all the time. I think maybe its fair to be explicit and call that out :D.

File: test/machine_test.go [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+		tmp.Host = elem.IP
+		tmp.Password = elem.Password
+		tmp.Username = elem.Username
+		tmp.Vendor = elem.Vendor
+
+		tests := map[string]expected{}
+		for _, elem := range elem.UseCases.Power {
+			t, ok := lookup[elem]
+			if !ok {
+				fmt.Printf("FYI, useCase: '%v' was not found. please verify it exists in the code base\n", elem)
```
Comment (mmlb) [inline test/machine_test.go:177]: shouldn't this panic or fail the test?

File: api/v1/task.proto (hunk 3 above)
Comment (mmlb) [reply api/v1/task.proto:38]: Can you split this change out into its own commit. All the other changes in this commit make sense in that they are needed to do the testing, but this one just seems to be non-obvious. Splitting to its own commit you can add that info in.

File: api/v1/task.proto [hunk 11]
```diff
@@ -4,7 +4,7 @@ option go_package = "github.com/tinkerbell/pbnj/api/v1";
 
 package github.com.tinkerbell.pbnj.api.v1;
 
-import "google/protobuf/any.proto";
+//import "google/protobuf/any.proto";
```
Comment (mmlb) [inline api/v1/task.proto:7]: delete?

File: .gitignore [hunk 12]
```diff
@@ -3,4 +3,5 @@ pbnj
 .idea*/**
```
Comment (mmlb) [inline .gitignore:3]: please you squash this commit into "add functional testing"

**Review by mmlb — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 12 above)
Comment (jacobweinstock) [reply .gitignore:3]: sorry, you're going to have to be more specific. 
image

File: test/machine_test.go (hunk 10 above)
Comment (jacobweinstock) [reply test/machine_test.go:177]: I believe It's a matter of preference. for me, if only one of multiple test use cases isn't found I think we should still run the found ones.

File: server/grpcsvr/taskrunner/taskrunner.go (hunk 9 above)
Comment (jacobweinstock) [reply server/grpcsvr/taskrunner/taskrunner.go:12]: you sure? this is how `go fmt` and `goimports` formatted this file.

File: test/README.md (hunk 5 above)
Comment (jacobweinstock) [reply test/README.md:48]: faking? no, this is for against real hardware

File: api/v1/task.proto (hunk 11 above)
Comment (jacobweinstock) [reply api/v1/task.proto:7]: I'll come back around after this is merged to clean up.

File: api/v1/task.proto (hunk 3 above)
Comment (jacobweinstock) [reply api/v1/task.proto:38]: we talked and I said I was going to go back to `google.protobuf.Any details` but I've changed my mind. time constraints and ease (creating a pb.any type is a lot more involved than just a string slice) of testing I'm going to leave it as is.

**Comment by jacobweinstock** [comment]
@mikemrm @jgavinray @mmlb  just a friendly ping :)

File: .gitignore (hunk 12 above)
Comment (mmlb) [reply .gitignore:3]: wow my grammar there :D
Squash 2a9a056 into 89de5bc1

File: api/v1/task.proto (hunk 3 above)
Comment (mmlb) [reply api/v1/task.proto:38]: heh. then lets go back and do this one change in its own commit with this info

File: cmd/machine.go [hunk 13]
```diff
@@ -32,10 +32,10 @@ var machineCmd = &cobra.Command{
 			os.Exit(1)
 		}
 		defer zlog.Sync() // nolint
-		logger.V(0).Info("machine actions")
+		//logger.V(0).Info("machine actions")
 
 		opts = append(opts, grpc.WithInsecure())
-		conn, err := grpc.Dial("localhost:50051", opts...)
+		conn, err := grpc.Dial("localhost:"+port, opts...)
```
Comment (mikemrm) [inline cmd/machine.go:37]: Is this to connect to the local pbnj daemon?

File: cmd/server.go [hunk 14]
```diff
@@ -80,5 +82,6 @@ var serverCmd = &cobra.Command{
 }
 
 func init() {
+	serverCmd.PersistentFlags().StringVar(&port, "port", "50051", "server port (default is 50051")
```
Comment (mikemrm) [inline cmd/server.go:68]: Will this conflict with #20 setting the default port to 9090?

File: test/machine_test.go [hunk 15]
```diff
@@ -122,14 +122,14 @@ func TestPower(t *testing.T) {
...
-func runClient(in testResource, action v1.PowerRequest_Action) (*v1.StatusResponse, error) {
+func runClient(in testResource, action v1.PowerRequest_Action, s Server) (*v1.StatusResponse, error) {
 	var opts []grpc.DialOption
 	ctx := context.Background()
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	opts = append(opts, grpc.WithInsecure())
-	conn, err := grpc.Dial("localhost:"+port, opts...)
+	conn, err := grpc.Dial(s.URL+":"+s.Port, opts...)
```
Comment (mikemrm) [inline test/machine_test.go:132]: this is only updated for the client and not the server right?

**Review by mikemrm — CHANGES_REQUESTED** [review]

File: cmd/machine.go (hunk 13 above)
Comment (jacobweinstock) [reply cmd/machine.go:37]: yes, will need to come back around and make `localhost` a flag too to allow for interaction with remote instances.

File: cmd/server.go (hunk 14 above)
Comment (jacobweinstock) [reply cmd/server.go:68]: yeah, i'll need to update this. I'd prefer to come back around and do that as this PR is already large and moving quite slow.

File: test/machine_test.go (hunk 15 above)
Comment (jacobweinstock) [reply test/machine_test.go:132]: yeah, only client here.

File: .gitignore (hunk 12 above)
Comment (jacobweinstock) [reply .gitignore:3]: updated

File: api/v1/task.proto (hunk 3 above)
Comment (jacobweinstock) [reply api/v1/task.proto:38]: updated

File: api/v1/task.proto (hunk 11 above)
Comment (jacobweinstock) [reply api/v1/task.proto]: updated

**Comment by jacobweinstock** [comment]
> Looks good, I think the intermediate drafts of test setup should all be squashed.

@mmlb mind having another look now that the commits are cleaned (hopefully) up?

**Review by mmlb — APPROVED** [review]
